modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/semantic/SemanticTermsEnricher.java

```diff
@@ -0,0 +1,96 @@
+import static org.mule.runtime.extension.api.connectivity.ConnectivityVocabulary.SECRET;
+import static org.mule.runtime.extension.api.util.ExtensionMetadataTypeUtils.getSemanticTerms;
+import static org.mule.runtime.extension.internal.semantic.SemanticTermsHelper.getTermsFromAnnotations;
+
+import org.mule.metadata.api.ClassTypeLoader;
+import org.mule.metadata.api.model.MetadataType;
+import org.mule.metadata.api.visitor.BasicTypeMetadataVisitor;
+import org.mule.runtime.api.meta.model.declaration.fluent.ConnectionProviderDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.OperationDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.ParameterDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.ParameterGroupDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.SourceDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.WithSemanticTermsDeclaration;
+import org.mule.runtime.extension.api.declaration.fluent.util.IdempotentDeclarationWalker;
+import org.mule.runtime.extension.api.declaration.type.ExtensionsTypeLoaderFactory;
+import org.mule.runtime.extension.api.loader.ExtensionLoadingContext;
+import org.mule.runtime.module.extension.api.loader.java.type.WithAnnotations;
+import org.mule.runtime.module.extension.internal.loader.enricher.AbstractAnnotatedDeclarationEnricher;
+import org.mule.runtime.module.extension.internal.loader.java.type.runtime.MethodWrapper;
+
+import java.util.Set;
+
+/**
+ * An enricher which adds semantic terms based on annotations
+ *
+ * @since 4.4.0ExtensionDocumentationResourceGenerator.java
+ */
+public class SemanticTermsEnricher extends AbstractAnnotatedDeclarationEnricher {
+
+  private ClassTypeLoader typeLoader = ExtensionsTypeLoaderFactory.getDefault().createTypeLoader();
+
+  @Override
+  public void enrich(ExtensionLoadingContext extensionLoadingContext) {
+    new IdempotentDeclarationWalker() {
+
+      @Override
+      protected void onConnectionProvider(ConnectionProviderDeclaration declaration) {
+        extractType(declaration).ifPresent(type -> addSemanticTerms(declaration, type));
+      }
+
+      @Override
+      protected void onOperation(OperationDeclaration declaration) {
+        extractImplementingMethod(declaration)
+            .map(method -> new MethodWrapper(method, typeLoader))
+            .ifPresent(method -> addSemanticTerms(declaration, method));
+      }
+
+      @Override
+      protected void onSource(SourceDeclaration declaration) {
+        extractType(declaration).ifPresent(type -> addSemanticTerms(declaration, type));
+      }
+
+      @Override
+      protected void onParameter(ParameterGroupDeclaration parameterGroup, ParameterDeclaration parameter) {
+        extractDeclaredParameter(parameter).ifPresent(e -> addSemanticTerms(parameter, e));
+        Set<String> typeTerms = getSemanticTerms(parameter.getType());
```

Comment:
The type annotation has an immutable set, I think that is good because that way you cannot mutate the annotation and therefore the metadataType that is annotated.

Is there something that I am missing, or will the add terms fail because it is immutable?

Reply:
good catch